File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode *slow , *fast;
        slow = fast = head;
        while(fast != NULL){
            if(fast->next !=NULL){
                if(fast->next->next != NULL){
                    fast = fast->next->next;
                }
                else{
                    return false;
                }
            }
            else{
                return false;
            }
            
            slow = slow->next;
            cout << "fast:" << fast->val << " slow:" << slow->val << endl;
            if(fast != NULL && slow != NULL){
                if(fast->val == slow->val){
                    return true;
                }
            }
            else{
                    return false;
                }
        }

        return false;
    }
};

File: Data Structures & Algorithms/maximum-twin-sum-of-a-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        int index = 0 , size = -1 , ans = 0;
        ListNode temp = head;
        do{
            ListNode temp2 = temp;
            int sum = 0;
            if(size == -1){
                int c = 1;
                while(temp2.next != null){
                    c+=1;
                    temp2 = temp2.next;
                }
                size = c;
                System.out.println("found size:"+size);
            }
            else{
                int i = index;
                while(i != (size - index -1)){
                    temp2 = temp2.next;
                    i+=1;
                }
                System.out.println(index+"<-- twin of -->"+ i );
            }
            sum = temp.val + temp2.val;
            index+=1;
            temp = temp.next;
            ans = Math.max(ans , sum);
        }while(index < size/2);
        return ans;
    }
}